### PR TITLE
[JENKINS-35412] fixed transitive plugin dependency resolution

### PR DESCRIPTION
**Pull request by daspilker**

### CODE REVIEW

**Comment by alexbrjo** [comment]
+1 I was having problems with Gradle dependencies and this seems to completely resolve them

**Comment by alexbrjo** [comment]
@daspilker what is the status on this? Are you waiting on reviews?

**Comment by abayer** [comment]
ping @daspilker

**Comment by mkobit** [comment]
Would love to see this get in.

**Comment by wolfs** [comment]
@daspilker Ping?

**Comment by wolfs** [comment]
@daspilker Yes, this seems like a workaround. I am not very happy with this change. My main objection not even the workaround, but that we resolve all the dependencies in configuration time just so that we can replace the `jpi` artifacts with `jar`s. I think there should be a better approach, but I don't know from the top of my head.
@oehme is this something which could be solved with variants?

**Comment by mkobit** [comment]
The way I handled the dependencies in https://github.com/mkobit/jenkins-pipeline-shared-libraries-gradle-plugin (which seems to work based on https://scans.gradle.com/s/mqma4cp63thiw ) was to use a `detachedConfiguration` and provide a `Callable` that only includes those files (see https://github.com/mkobit/jenkins-pipeline-shared-libraries-gradle-plugin/blob/d525c58cfc21cf9caee0ea7f887ba43a8fad6d5f/src/main/kotlin/com/mkobit/jenkins/pipelines/SharedLibraryPlugin.kt#L105-L121). You don't get the same output in `./gradlew dependencies` or anything but it does seem to make the resolution happen during execution rather than configuration.

I was planning on filing a feature request to Gradle around providing more metadata to dependency resolution APIs unless variants solves this (I don't know anything about variants).

**Comment by oehme** [comment]
You definitely shouldn't do this at configuration time. A better approach is to use `Configuration.incoming.beforeResolve`. In there you can create a copy of the configuration, resolve that copy and then mutate the original one before it is resolved. Not pretty, but at least not a configuration time performance breakdown.

Yes, we want to replace this with a variant query API, which would supersede the current ArtifactQuery API. The latter one only supports a hardcoded set of artifact types unfortunately.

**Comment by daspilker** [comment]
@wolfs Re https://github.com/jenkinsci/gradle-jpi-plugin/pull/92#issuecomment-335031179: I pushed another commit to not rely on a hard-coded `compileClassplath`. The `@jar` dependencies will be added when the first configuration is resolved where the hierarchy contains the target configuration. I tested this with job-dsl-plugin. Everything seems to be working, including listing dependencies and importing the project in IDEA.

File: src/main/groovy/org/jenkinsci/gradle/plugins/jpi/JpiPlugin.groovy [hunk 1]
```diff
@@ -349,4 +346,38 @@ class JpiPlugin implements Plugin<Project> {
...
+        ConfigurationContainer configurations = project.configurations
+        Configuration fromConfiguration = configurations.getByName(from)
+        Configuration toConfiguration = configurations.getByName(to)
+
+        project.configurations.each { Configuration c ->
+            c.incoming.beforeResolve { ResolvableDependencies resolvableDependencies ->
+                if (c.hierarchy.contains(toConfiguration) && fromConfiguration.state == UNRESOLVED) {
+                    fromConfiguration.resolvedConfiguration.resolvedArtifacts
+                            .findAll { it.type == 'hpi' || it.type == 'jpi' }
+                            .each { project.dependencies.add(to, "${it.moduleVersion.id}@jar") }
```
Comment (wolfs) [inline src/main/groovy/org/jenkinsci/gradle/plugins/jpi/JpiPlugin.groovy]: IIUC, we then have the `jpi` dependency and the `jar` dependency on the `to` configuration, given that these configurations extend from the `from` configurations. Is that what we really want? Shouldn't we better have only the `jar`s in there?

Comment (daspilker) [reply src/main/groovy/org/jenkinsci/gradle/plugins/jpi/JpiPlugin.groovy]: Good point. When I remote the calls to `extendsFrom`, job-dsl-plugin has test failures due to `java.lang.ClassNotFoundException`. Not sure why, I need to compare the dependencies. But does it hurt to have `jpi` and `jar` dependencies in the same configuration?

Comment (wolfs) [reply src/main/groovy/org/jenkinsci/gradle/plugins/jpi/JpiPlugin.groovy]: I guess it does not hurt. I think Java 9 has problems with files on the classpath which are not zips, but that is also not problematic here. I think the `ClassNotFoundExceptions` come from the fact that we only add the `jpi`/`hpi` artifacts here, but not the regular `jar` artifacts in the configurations. So I guess it should be something like:
```
fromConfiguration.resolvedConfiguration.resolvedArtifacts
    .each {
        if (it.type == 'hpi' || it.type == 'jpi') {
            project.dependencies.add(to, "${it.moduleVersion.id}@jar")
        } else {
            project.dependencies.add(to, it)
        }
    }
```

Comment (daspilker) [reply src/main/groovy/org/jenkinsci/gradle/plugins/jpi/JpiPlugin.groovy]: Ah, you are right. That was missing.

As it does not seem to be problematic, I would merge this PR as it is and deal with separating `jar` and `jpi` configurations in another PR. The `jenkinsTest` configuration currently contains both, the plugins for testing (`jpi`/`hpi' ) and the Jenkins Test Harness (`jar`), so we would need to split that to get a clean separation.

I'll file a JIRA issue for separating the configuration, merge this PR and cut a release if there are no objections.

Comment (wolfs) [reply src/main/groovy/org/jenkinsci/gradle/plugins/jpi/JpiPlugin.groovy]: 👍 SGTM!

Comment (daspilker) [reply src/main/groovy/org/jenkinsci/gradle/plugins/jpi/JpiPlugin.groovy]: The JIRA issue for separating the configurations: https://issues.jenkins-ci.org/browse/JENKINS-47375

**Comment by wolfs** [comment]
@daspilker I guess this solution is about the same as we are doing now: https://discuss.gradle.org/t/configuration-and-transitive-dependency-management-by-extension/23824/5

**Pull request merged**